### PR TITLE
V2 APIs

### PR DESCRIPTION
**Pull request by sooraj1002**
the implementation for v2 APIs tried to keep the core logic of the APIs same as earlier, with the business logic of iteration and generation using the same functions. 

along with that, it added more control within the API itself in terms of gevent concurrency, batch sizes for examples and maximum iterations.

it also adds new features: 
- user examples are validated against the model defined in the config, using pydantic models created with https://github.com/BharatSahAIyak/autotune/pull/43
- mapping of generated examples against prompt, so that a user can check which prompt worked best for them.
- Everything costs money, and now, we track and monitor our spending on those precious credits!
- To reduce costs with lesser db calls, used django's built in cache for getting the user info in f17564ac88bc466e63edfb2e45704de1583d1e9c

Bug fixes:
- while adding user examples to the prompt, only add those examples which have a reason associated with them [here](https://github.com/BharatSahAIyak/autotune/blob/bb1b896bb52ef90333ae2c1d0a5b7a26078a8c85/workflow/dataFetcher.py#L149C1-L152C14)
- fixed a bug where sometimes examples arent returned in iteration by storing generated examples with class objects, instead of recursively returning them.
- fixed bug where the number of generated examples was getting overwritten in the db while generating. this happened because we werent updating the task object against the db in `celery_task.py` after generating.

fix:
https://github.com/BharatSahAIyak/autotune/blob/7b16ca0d328addb3a08b7f3a5b4611f4ec59c773/workflow/celery_task.py#L45

the bug was exposed and fixed while creating the /status endpoint

### CODE REVIEW

File: autotune/urls.py
Comment (KDwevedi) [inline autotune/urls.py:1]: If we're versioning APIs and supporting multiple versions, we should use namespaces and segregate them like so:

```python
path("api/v1/workflow/", include(("workflow.urls", 'workflow'), namespace='v1'))
path("api/v2/workflow/", include(("workflowV2.urls", 'workflowV2'), namespace='v2'))
```

File: docs/API v2/GET status.md
Comment (KDwevedi) [inline docs/API v2/GET status.md:1]: Let's standardise how we name components of the API

**Headers: First letter of each word captal, separating multiple words by hyphens**
> My-Header: Some Value

**query params: smallcase all the way through, separating words by hyphens
> https://my-url.com?query-paramenter-name=queryValue

File: autotune/urls.py
Comment (sooraj1002) [reply autotune/urls.py:1]: v2 is just to differentiate the API we are providing NISAI. it is basically a consolidated version of the original APIs.

File: docs/API v2/GET status.md
Comment (sooraj1002) [reply docs/API v2/GET status.md:1]: django rest framework handles this for us. it converts all request headers to the format you have specified. 

made the changes to the query params. https://github.com/BharatSahAIyak/autotune/pull/48/commits/ffc74d7a476bbf73ab06369af04e46e8f16ee461

File: autotune/urls.py
Comment (sooraj1002) [reply autotune/urls.py:1]: added versioning https://github.com/BharatSahAIyak/autotune/pull/48/commits/4d930427a1a5f5960ffe173e5102e307e7c7d27f

File: workflow/celery_task.py [hunk 1]
```diff
@@ -27,17 +28,39 @@ def process_task(self, task_id):
...
         workflow_config_id=workflow.workflow_config.id,
         llm_model=workflow.llm_model,
+        Model=Model,
         refine=True,
         task_id=task_id,
         iteration=1,
     )
 
+    task.refresh_from_db()
+    print(f"generated samples= {task.generated_samples} in celery")
```
Comment (KDwevedi) [inline workflow/celery_task.py:50]: Is this print statement intended?

Comment (KDwevedi) [inline workflow/celery_task.py:50]: Is this print statement intended?

Comment (sooraj1002) [reply workflow/celery_task.py:50]: Helps while debugging

File: workflow/celery_task.py [hunk 2]
```diff
@@ -27,17 +28,39 @@ def process_task(self, task_id):
...
     task.save()
 
-    fetcher = DataFetcher()
+    Model, _ = create_pydantic_model(workflow.workflow_config.schema_example)
+
+    fetcher = DataFetcher(
+        max_iterations=max_iterations,
+        max_concurrent_fetches=max_concurrent_fetches,
+        batch_size=batch_size,
+    )
```
Comment (KDwevedi) [inline workflow/celery_task.py:37]: Are values for `max_iterations`, `batch_size` etc getting validated anywhere?
Can -ve values be pushed here by API calls?

File: workflow/dataFetcher.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        workflow_id,
+        total_examples,
+        workflow_config_id,
+        llm_model,
+        Model,
+        refine=False,
+        task_id=None,
+        iteration=None,
+    ):
+        if iteration is not None and iteration > self.max_iterations:
```
Comment (KDwevedi) [inline workflow/dataFetcher.py:41]: What if iteration *is* `None`?

File: workflow/dataFetcher.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+            logger.error("Max iterations reached")
+            return
+        user_prompt, prompt_id = self.construct_user_prompt(workflow_id, refine)
+        logger.info(workflow_config_id)
+        config = get_object_or_404(WorkflowConfig, id=workflow_config_id)
+        try:
+            total_batches = max(
+                1,
+                (total_examples - self.generated + self.batch_size - 1)
+                // self.batch_size,
```
Comment (KDwevedi) [inline workflow/dataFetcher.py:51]: Unsafe integer division?

File: workflow/dataFetcher.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+                    prompt_id,
+                )
+                for batch_index in range(
+                    min(total_batches, self.max_concurrent_fetches)
+                )
+            ]
+
+            joinall(greenlets)
+
+            if task_id is not None:
```
Comment (KDwevedi) [inline workflow/dataFetcher.py:75]: Can Task ID be `None` and the if below still be executed?
Why not handle `else` here as well?

File: workflow/dataFetcher.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        fields,
+        prompt_id,
+    ):
+        response = self.call_llm_generate(
+            user_prompt, workflow_config_id, llm_model, iteration, batch_index
+        )
+
+        logger.info("response received from LLM")
+
+        if response:
```
Comment (KDwevedi) [inline workflow/dataFetcher.py:126]: What happens if response fails or isn't truth-ey?

File: workflow/celery_task.py (hunk 2 above)
Comment (sooraj1002) [reply workflow/celery_task.py:37]: This does not have any checks right now for malicious users.

File: workflow/dataFetcher.py (hunk 3 above)
Comment (sooraj1002) [reply workflow/dataFetcher.py:41]: then only one call to openAI will be made, and the result to the user will be based on that call.

**Review by KDwevedi — APPROVED** [review]
LGTM

**Pull request merged**